contributors/design-proposals/storage/container-storage-interface-cluster-registry.md

```diff
@@ -0,0 +1,145 @@
+* [CSI Volume Plugins in Kubernetes Design Doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md)
+
+## Objective
+
+The objective of this document is to document all the requirements for enabling a cluster level registry for CSI compliant volume plugins (a CSI volume driver) in Kubernetes.
+
+## Goals
+
+* Allow a CSI volume plugin (that opts-in to new mechanism) to configure how Kubernetes should interact with it.
+* Improve discoverability by users/cluster admins of a CSI volume plugin (that opts-in to new mechanism) deployed on the Kubernetes cluster.
+
+## Non-Goals
+
+* Define how CSI volume plugin should be deployed on Kubernetes.
+* Require use of new CSI cluster registry mechanism for all CSI volume plugins deployed on Kubernetes (the mechanism will be optional/opt-in).
+
+## Design Overview
+
+A new custom resource will be automatically be installed on Kubernetes clusters.
+
+Upon deployment a plugin must create a new custom resource object.
+
+## Design Details
+
+### `CSIPlugin` Object
+
+#### Proposed API
+
+```go
+// CSIPlugin captures information about a Container Storage Interface (CSI)
+// volume plugin deployed on the cluster.
+//
+// CSIPlugin objects are non-namespaced.
+type CSIPlugin struct {
+    metav1.TypeMeta   `json:",inline"`
+
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+    metav1.ObjectMeta `json:"metadata"`
+
+    // Specification describing the CSI volume plugin and any custom
+    // configuration for it.
+	Spec CSIPluginSpec `json:"spec"`
+
+	// Status of the CSI volume plugin.
+	Status CSIPluginStatus `json:"status,omitempty"`
+}
+
+// CSIPluginList is a collection of CSIPlugin objects.
+type CSIPluginList struct {
+    metav1.TypeMeta `json:",inline"`
+    
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is the list of CSIPlugins
+	Items []CSIPlugin `json:"items"`
+}
+
+// CSIPluginSpec is the specification of a CSIPlugin.
+type CSIPluginSpec struct {
+    // Driver indicates the name of the CSI Plugin that this object refers to.
+    // This MUST be the same name returned by the CSI GetPluginName() call for
+    // that driver.
+    Driver string `json:"driver"`
+    
+    // Indicates this CSI volume plugin does not require an attach operation
+    // (because it does not implement the ControllerPublishVolume() method),
+    // and that Kubernetes should not call or wait for any attach operation and
+    // just skip to mounting.
+	SkipAttach bool `json:"skipAttach"`
+}
+
+// CSIPluginStatus is the status of a CSIPlugin.
+type CSIPluginStatus struct {
+    // Indicates the volume plugin has been successfully deployed on the cluster
+    // and is ready to use.
+	Ready bool `json:"ready"`
```

Comment:
I'm hesitant on expanding the API more for a hypothetical use case. I'd prefer to leave it simple or drop it altogether until we have a real use case. If I drop it, there would be nothing in `CSIPluginStatus` so I would have to drop that for now too. What do you prefer?

Reply:
I am torn on this. Currently I don't see any useful Status to report. But I may come in future. Can we have an object with .Spec and no .Status?

If we use `Ready`, we need clear semantics what it means and most importantly what is it good for. Is it enough to have one node with CSI driver installed to have the global object ready? Is it enough to have one node with CSI driver broken (failed liveness probe, driver not installed there, crash loop...) to have the global object not-ready?

>  In the future we can use this as one of the bits for smarter CSI volume aware scheduling.

For smarter CSI volume aware scheduling I would use decision based on availability / status of drivers on nodes, not a global object. What should scheduler do when the global object is not ready? All it can do is to stop scheduling.